server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java

```diff
@@ -145,12 +178,12 @@ interface Leaf {
             /**
              * Position the loader at a document.
              */
-            boolean advanceToDoc(int docId) throws IOException;
+            boolean advanceToDoc(FieldsVisitor fieldsVisitor, int docId) throws IOException;
 
             /**
              * Write values for this document.
              */
-            void write(XContentBuilder b) throws IOException;
+            void write(FieldsVisitor fieldsVisitor, XContentBuilder b) throws IOException;
```

Comment:
I don't think I actually need `fieldsVisitor` here - I think `advanceToDoc` can grab it.

Reply:
Yep. I wonder if we can avoid having it as a parameter in any of these methods and instead pass it StoredFieldSourceLoader implementations directly? Having a method param that is only used by a specific subset of implementations feels off to me.

Reply:
Like in the ctor?

Reply:
I could move it to the `leaf` method pretty easily. But it's kind of tricky because you have to advance the state in a specific way. And holding on to a reference to the thing for a while feels like it is more "at a distance". Like, we take a `docId` as a parameter, but we only use it if we're using doc values.